notebooks/structured/credit_card_fraud.livemd

```diff
@@ -1,141 +1,176 @@
 
 ```elixir
-# Download data with a Kaggle account: https://www.kaggle.com/mlg-ulb/creditcardfraud/
-@file_name "creditcard.csv"
-
-defp data() do
-  IO.puts("Loading #{@file_name}")
-  df = Explorer.DataFrame.from_csv!(@file_name, dtypes: [{"Time", :float}])
+data_path_input = Kino.Input.text("Data path (CSV)")
+```
 
-  {train_df, test_df} = split_train_test(df, 0.8)
+Now, let's read the data into an `Explorer.Dataframe`:
 
-  IO.puts("Training Samples: #{inspect(Explorer.DataFrame.n_rows(train_df))}")
-  IO.puts("Testing Samples: #{inspect(Explorer.DataFrame.n_rows(test_df))}")
-  IO.write("\n\n")
+```elixir
+data_path = Kino.Input.read(data_path_input)
 
-  {train_features, train_targets} = split_features_targets(train_df)
-  {test_features, test_targets} = split_features_targets(test_df)
+df = DataFrame.from_csv!(data_path, dtypes: [{"Time", :float}])
+```
 
-  train_features = normalize_data(train_features)
-  test_features = normalize_data(test_features)
+For further processing, we will need a couple helper functions. We will group them in a module for convenience.
 
-  {
-    {df_to_tensor(train_features), df_to_tensor(train_targets)},
-    {df_to_tensor(test_features), df_to_tensor(test_targets)}
-  }
-end
+```elixir
+defmodule CredidCard.Data do
+  import Nx.Defn
 
-defp split_train_test(df, portion) do
-  num_examples = Explorer.DataFrame.n_rows(df)
-  num_train = ceil(portion * num_examples)
-  num_test = num_examples - num_train
+  def split_train_test(df, portion) do
+    num_examples = DataFrame.n_rows(df)
+    num_train = ceil(portion * num_examples)
+    num_test = num_examples - num_train
 
-  {
-    Explorer.DataFrame.slice(df, 0, num_train),
-    Explorer.DataFrame.slice(df, num_train, num_test)
-  }
-end
+    train = DataFrame.slice(df, 0, num_train)
+    test = DataFrame.slice(df, num_train, num_test)
+    {train, test}
+  end
 
-defp split_features_targets(df) do
-  features = Explorer.DataFrame.select(df, &(&1 == "Class"), :drop)
-  targets = Explorer.DataFrame.select(df, &(&1 == "Class"), :keep)
-  {features, targets}
-end
+  def split_features_targets(df) do
+    features = DataFrame.select(df, &(&1 == "Class"), :drop)
+    targets = DataFrame.select(df, &(&1 == "Class"), :keep)
+    {features, targets}
+  end
 
-defp normalize(name),
-  do: fn df ->
-    Explorer.Series.divide(
-      df[name],
-      Explorer.Series.max(
-        Explorer.Series.transform(df[name], fn x ->
-          if x >= 0 do
-            x
-          else
-            -x
-          end
-        end)
-      )
-    )
+  def df_to_tensor(df) do
+    df
+    |> DataFrame.names()
+    |> Enum.map(&Series.to_tensor(df[&1]))
+    |> Nx.stack(axis: 1)
   end
 
-defp normalize_data(df) do
-  df
-  |> Explorer.DataFrame.names()
-  |> Map.new(&{&1, normalize(&1)})
-  |> then(&Explorer.DataFrame.mutate(df, &1))
-end
+  defn normalize_features(tensor) do
```

Comment:
Perhaps we can use Scholar here and demo Axon+Scholar in tandem?

Reply:
Definitely, I would wait for a Scholar release though. The other functions look common enough to also have _somewhere_ eventually.